python/mxnet/amp/amp.py

```diff
@@ -0,0 +1,248 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# coding: utf-8
+"""Functions for enabling AMP (automatic mixed precision)."""
+__all__ = ['init']
+
+import logging
+import contextlib
+import numpy as np
+from types import MethodType
+from functools import partial
+
+from .. import symbol
+from ..symbol import Symbol
+from .. import ndarray
+from ..ndarray import NDArray
+from . import lists
+from ..gluon import trainer
+from .. import optimizer as opt
+from .loss_scaler import *
+
+def _cast_symbol_NDArray(s, dtype):
+    if isinstance(s, Symbol):
+        return symbol.amp_cast(s, dtype=dtype)
+    elif isinstance(s, NDArray):
+        if s.dtype != dtype and (s.dtype == np.float16 or s.dtype == np.float32):
+            return ndarray.amp_cast(s, dtype=dtype)
+        else:
+            return s
+    else:
+        return s
+
+def _wrap_symbol_functions(module):
+    def _ndarray_wrapper(f, target_dtype, cond_arg=None):
+        def _new_fun(*args, **kwargs):
+            if cond_arg is not None:
+                if (cond_arg[0] not in kwargs or
+                        kwargs[cond_arg[0]] not in cond_arg[1]):
+                    return f(*args, **kwargs)
+            new_args = list(map(lambda x: _cast_symbol_NDArray(x, target_dtype), args))
```

Comment:
If an fp16 output is used for multiple fp32 operators, can we cast the fp16 only once?

Reply:
That would require a graph pass approach instead of simple function substitution.

Reply:
This will degrade the performance a lot so I suggest using the graph pass.

Reply:
Adding a graph pass may be the next step for performance tuning but it definitely is not necessary for the feature to be useful. I tested all the models from GluonCV and in none of them I saw a need for it (the unnecessary casts typically take <1% of the total time).

Reply:
Sounds good :)  Could you share your data and the plan as my comments in the below?